pennylane/interfaces/batch/jax.py

```diff
@@ -0,0 +1,257 @@
+                non_diff_wrapper,
+                args,
+                result_shape=jax.ShapeDtypeStruct((total_params,), dtype),
+            )
+
+            param_idx = 0
+            res = []
+
+            # Group the vjps based on the parameters of the tapes
+            for p in params:
+                res.append(vjps[param_idx : param_idx + len(p)])
+                param_idx += len(p)
+
+                # Unpack if we have the vjp for a single parameter
+                if len(p) == 1:
+                    res[-1] = res[-1][0]
+
+            # Unwrap partial results into arrays if need be
+            need_unwrapping = any(r.ndim != 0 for r in res)
+            if need_unwrapping:
+                unwrapped_res = []
+                for r in res:
+                    if r.ndim != 0:
+                        r = [jnp.array(p) for p in r]
+                    unwrapped_res.append(r)
+
+                res = unwrapped_res
+
+            return (tuple(res),)
+
+        # Gradient function is a device method.
+        with qml.tape.Unwrap(*tapes):
+            jacs = gradient_fn(tapes, **gradient_kwargs)
+
+        vjps = [qml.gradients.compute_vjp(d, jac) for d, jac in zip(g, jacs)]
+        res = [[jnp.array(p) for p in v] for v in vjps]
+        return (tuple(res),)
+
+    wrapped_exec.defvjp(wrapped_exec_fwd, wrapped_exec_bwd)
+    return wrapped_exec(params)
+
+
+# The execute function in forward mode
+def _execute_with_fwd(
+    params,
+    tapes=None,
+    device=None,
+    execute_fn=None,
+    gradient_kwargs=None,
+    _n=1,
+):  # pylint: disable=dangerous-default-value,unused-argument
+
+    # Only have scalar outputs
+    total_size = len(tapes)
+
+    @jax.custom_vjp
+    def wrapped_exec(params):
+        def wrapper(p):
+            """Compute the forward pass by returning the jacobian too."""
+            new_tapes = []
+
+            for t, a in zip(tapes, p):
+                new_tapes.append(t.copy(copy_operations=True))
+                new_tapes[-1].set_parameters(a)
+
+            res, jacs = execute_fn(new_tapes, **gradient_kwargs)
+
+            # On the forward execution return the jacobian too
+            return np.stack(res), jacs
+
+        jacobian_shape = [jax.ShapeDtypeStruct((1, len(p)), dtype) for p in params]
+        res, jacs = host_callback.call(
+            wrapper,
+            params,
+            result_shape=tuple([jax.ShapeDtypeStruct((total_size, 1), dtype), jacobian_shape]),
+        )
+        return res, jacs
+
+    def wrapped_exec_fwd(params):
+        res, jacs = wrapped_exec(params)
+        return res, tuple([jacs, params])
+
+    def wrapped_exec_bwd(params, g):
+
+        # Use the jacobian that was computed on the forward pass
+        jacs, params = params
+        res_jacs = []
+        for j in jacs:
+            this_j = []
+            for i in range(j.shape[1]):
+                this_j.append(j[0, i])
+            res_jacs.append(this_j)

```

Comment:
What is happening here?

Reply:
We're creating the same nested structure that is being created in non-forward mode:
```python
[DeviceArray([[ 0.06695931,  0.01383095, -0.46500877]], dtype=float32)]
```
is mapped to
```python
[[DeviceArray(0.06695931, dtype=float32), DeviceArray(0.01383095, dtype=float32), DeviceArray(-0.46500877, dtype=float32)]]
```
This is to ensure that all modes return the same structure for the jacobian.

Reply:
I would also add this as a comment to the code!